### PR TITLE
Make run and rmi bash completions configurable

### LINKED ISSUE #9474: Bash completion is slow for `docker run`, `docker rmi`
# In detail (see below for short version)
When doing tab completion for image names for the following commands, there is always a 1 - 2 seconds delay after pressing tab:
* `docker run <tab>`
* `docker rmi <tab>`
* ...

Additionally, a large number of possible completions is generated (389 on my system), consisting of repositories, image names and image ids. Most of these (~350) are intermediate image ids, which should almost never be used by end users anyway.

I'd also argue that there is no real use case for completing image ids. Normal users won't normally be using numeric image ids, except for debugging. If I don't know the image id, seeing multiple image ids with the same prefix won't help me. (The only use case I can think of are sysadmins).

# TL; DR
There are two problems here:
* Slow completion (1 - 2 seconds is definitely to slow for interactive use)
* Too many (non-useful) completions (389 is too much)

I'd argue that there are the following options to fight the problems above:
* Show everything, including intermediate image ids (like `docker images --all`). This is the current situation.
* Show non-intermediate image ids (like `docker images`).
* Only show repositories and tags.
* Only show repositories, and show tags only after the user has typed `:`

I'd vote for making this configurable (e.g. via an environment variable in `.bashrc` or globally via another file), and making the third option the default. The delay as well as the number of generated completions decrease from top to bottom.

# Addendum: Proposal
The configuration option could look like this:
```bash
DOCKER_IMAGE_COMPLETION=(all_images|non_intermediate_images),repositories,tags
```
the default would be:
```bash
DOCKER_IMAGE_COMPLETION=repositories
```

### PR DESCRIPTION
**Pull request by pugnascotia**
Allow the user to configure how Docker's bash completion works for the "run" and "rmi" commands through the following environment variables:

DOCKER_COMPLETION_SHOW_IMAGE_IDS
- "none" - Show names only (default)
- "non-intermediate" - Show names and ids, but omit intermediate image IDs
- "all" - Show names and ids, including intermediate image IDs

DOCKER_COMPLETION_SHOW_TAGS
- "yes" - include tags in completion options
- "no"  - don't include tags in completion options (default)

Note that I haven't changed the zsh completion - it works quite differently and I think it's fine as it is.

Fixes #9474

### CODE REVIEW

**Comment by pugnascotia** [comment]
Sorry about that, should be signed now

**Comment by pugnascotia** [comment]
...I don't think the windows build failure is my fault?

**Comment by thaJeztah** [comment]
@pugnascotia nah, don't think so. No worries :D

File: contrib/completion/bash/docker [hunk 1]
```diff
@@ -70,9 +85,67 @@ __docker_container_ids() {
 	COMPREPLY=( $(compgen -W "${containers[*]}" -- "$cur") )
 }
 
+__docker_image_generate_completions() {
+	local counter=$( __docker_pos_first_nonflag $( __docker_to_alternatives "$options_with_args" ) )
+
+	if [ $cword -eq $counter ]; then
+
```
Comment (albers) [inline contrib/completion/bash/docker]: I think position checking should not be moved into this function.
It's a different concern and also introduces a coupling to the number of images that can be completed (just one).

File: contrib/completion/bash/docker [hunk 2]
```diff
@@ -70,9 +85,67 @@ __docker_container_ids() {
...
+		echo "$cur" | grep -q ':$'
+		if [ $? -eq 0 ]; then
+			__docker_image_repos_and_tags
+		else
+			case "${DOCKER_COMPLETION_SHOW_IMAGE_IDS:-"none"}" in
+				all)
+					case "${DOCKER_COMPLETION_SHOW_TAGS:-"no"}" in
+						yes)
+							__docker_image_repos_and_tags_ids_with_intermediates
```
Comment (albers) [inline contrib/completion/bash/docker]: This function exists under a different name.
Corresponding execution path errors out with

    __docker_image_repos_and_tags_ids_with_intermediates: command not found

**Comment by pugnascotia** [comment]
Fixes push - good catch

**Comment by albers** [comment]
With

    DOCKER_COMPLETION_SHOW_IMAGE_IDS=non-intermediate
    DOCKER_COMPLETION_SHOW_TAGS=yes

`docker run`also completes intermediate images.

**Comment by albers** [comment]
There is some repetition in the individual completion functions.
Maybe the whole business could be handled in one function that assembles the docker and awk commands. Or the awk script could be configurable.
Can you please give it a try?

**Comment by albers** [comment]
By the way: 
Thanks for taking care of this completion issue, and thanks for the contribution!
It's really nice and I'm very much looking forward to having this functionality available.
I like the defaults.

**Comment by albers** [comment]
@pugnascotia take a look at how the docker command is assembled in `__docker_q`. Maybe this could be helpful here.

**Comment by pugnascotia** [comment]
I've centralised the changes into __docker_images() and revert some other changes.

File: contrib/completion/bash/docker [hunk 3]
```diff
@@ -70,6 +85,57 @@ __docker_container_ids() {
...
+			images_args="--no-trunc -a"
+			;;
+		non-intermediate)
+			images_args="--no-trunc"
+			;;
+	esac
+
+	local awk_script=''
+
+	case "$DOCKER_COMPLETION_SHOW_IMAGE_ID" in
```
Comment (albers) [inline contrib/completion/bash/docker]: Should be `$DOCKER_COMPLETION_SHOW_IMAGE_IDS`.
Currently `DOCKER_COMPLETION_SHOW_IMAGE_IDS` has no effect at all.

File: contrib/completion/bash/docker [hunk 4]
```diff
@@ -70,6 +85,57 @@ __docker_container_ids() {
...
+		all)
```
Comment (albers) [inline contrib/completion/bash/docker]: As the `all` and `non-intermediate` branches are identical, they should be combined.

Comment (pugnascotia) [inline contrib/completion/bash/docker]: This is what happens when you knock out a change at 5PM on a Friday. I'll
get that fixed up on Monday. Thanks for the feedback!

On 8 August 2015 at 13:27, Harald Albers <notifications@github.com> wrote:

> In contrib/completion/bash/docker
> <https://github.com/docker/docker/pull/15332#discussion_r36577802>:
>
> > @@ -70,6 +85,57 @@ __docker_container_ids() {
> >  	COMPREPLY=( $(compgen -W "${containers[*]}" -- "$cur") )
> >  }
> >
> > +__docker_images() {
> > +	local images_args=""
> > +	case "$DOCKER_COMPLETION_SHOW_IMAGE_IDS" in
> > +		all)
> > +			images_args="--no-trunc -a"
> > +			;;
> > +		non-intermediate)
> > +			images_args="--no-trunc"
> > +			;;
> > +	esac
> > +
> > +	local awk_script=''
> > +
> > +	case "$DOCKER_COMPLETION_SHOW_IMAGE_ID" in
> > +		all)
>
> As the all and non-intermediate branches are identical, they should be
> combined.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/15332/files#r36577802>.
>



-- 
Rory Hunter

**Comment by pugnascotia** [comment]
Merged those cases and fixed the variable name.

File: contrib/completion/bash/docker [hunk 5]
```diff
@@ -70,6 +85,46 @@ __docker_container_ids() {
...
+			images_args="--no-trunc"
+			;;
+	esac
+
+	local awk_script=''
+	case "$DOCKER_COMPLETION_SHOW_IMAGE_IDS" in
+		all|non-intermediate)
+			case "$DOCKER_COMPLETION_SHOW_TAGS" in
+				yes)
+					awk_script='NR>1 { print $3; if ($1 != "<none>") { print $1; print $1":"$2 } }'
```
Comment (albers) [inline contrib/completion/bash/docker]: Thanks for the update, Rory. I found no problems any more.

But I suggest one more refactoring: If you pull out the part that prints the repo names (`print $1; print $1":"$2` vs `print $1`) from `awk_script` into a new variable and insert its value here, you get rid of some repitition and reduce nesting. What do you think?

**Comment by pugnascotia** [comment]
I've had a bash (ho ho) at this, see what you think.

**Comment by albers** [comment]
I really like how you express the final `awk` script. This reads very well.

But the code still preserves the nested structure (with the `if` inside the `case`) and therefore IMHO is not a real improvement.

What I was aiming at is that - in the previous state of the code - you could replace both `case "$DOCKER_COMPLETION_SHOW_TAGS" in` blocks with just one line, e.g.

    awk_script='NR>1 { print $3; if ($1 != "<none>") { '"$repo_print_command"' } }'

and

    awk_script='NR>1 && $1 != "<none>" { '"$repo_print_command"' }'    

if you factor out just the difference in the awk print command to a new variable.

**Comment by pugnascotia** [comment]
Ah, I see what you mean. Try the latest changes.

**Comment by albers** [comment]
Well, actually I was thinking more of 
```
        local repo_print_command
        if [ "$DOCKER_COMPLETION_SHOW_TAGS" = "yes" ]; then
                repo_print_command='print $1; print $1":"$2'
        else
                repo_print_command='print $1'
        fi

        local awk_script
        case "$DOCKER_COMPLETION_SHOW_IMAGE_IDS" in
                all|non-intermediate)
                        awk_script='NR>1 { print $3; if ($1 != "<none>") { '"$repo_print_command"' } }'
                        ;;
                none|*)
                        awk_script='NR>1 && $1 != "<none>" { '"$repo_print_command"' }'
                        ;;
        esac
```

**Comment by pugnascotia** [comment]
That looks simpler. Updates pushed.

**Comment by albers** [comment]
It works like a charm and the code is OK. Thanks for your patience, Rory. It's not easy to talk about code changes.

Is there a specific reason why you only added this feature to the `run`, `create` and `rmi` subcommands?

For consistency I suggest we also use the new `__docker_images` function in the remaining occurrances of `__docker_image_repos_and_tags_and_ids`, i.e.
* `docker events -f image=`
* `docker history`
* `docker inspect`
* `docker save`
* in `__docker_containers_and_images()`

**Comment by albers** [comment]
There's another point I'd like to discuss:

After some fiddling with the new variants, I am no longer convinced that the default value for `DOCKER_COMPLETION_SHOW_TAGS` of `no` is a good choice.
After all, it means you get no support for completing tags at all, which turns out to be confusing.
Any opinions about this?

**Comment by pugnascotia** [comment]
I originally modified only the `rm` and `rmi` completions as they were specifically mentioned on the issue, but I've now expanded it to the other completions since `__docker_image_repos_and_tags_and_ids`  function was just duplicating `__docker_images`.

I've also defaulted tags completion to `"yes"` - you're right, it would be confusing if tags didn't complete by default. They can now be disabled if someone so chooses.

**Comment by albers** [comment]
Perfect! LGTM.

ping @tianon @jfrazelle

**Comment by pugnascotia** [comment]
...do I need to do anything further? First contribution, you see :-)​

**Comment by thaJeztah** [comment]
@pugnascotia you did great so far! thanks so much for contributing :+1: 

We prefer to have the commits squashed in logical units. For most PR's that means a single commit (also see [contributing.md](https://github.com/docker/docker/blob/master/CONTRIBUTING.md#conventions)

If you need help with squashing, then [this section in our documentation](https://docs.docker.com/project/work-issue/#pull-and-rebase-frequently) could be helpful. For more information on the contributing process, the [Open Source at Docker](https://docs.docker.com/opensource/how-to-contribute/) docs contains a lot of useful information.

For this PR, I think squashing down to a single commit would be good.

Next steps will be code review (if still needed after @albers excellent work), and, possibly docs review. (See the [pull request reviewing process](https://github.com/docker/docker/blob/master/project/REVIEWING.md))

(I don't think we currently have documentation on the completion scripts for Docker itself, I see *compose* does have some docs on [command completion](https://docs.docker.com/compose/completion/), maybe docker should have something as well?)

**Comment by thaJeztah** [comment]
WRT the documentation; Adding such document will be out of scope for this PR, just thinking out loud where/how we should document this feature :smile:

**Comment by pugnascotia** [comment]
Rebased and squashed.

**Comment by thaJeztah** [comment]
ping @tianon @jfrazelle for review :heart:

**Comment by jfrazelle** [comment]
this is super cool, thanks @albers for reviewing as well

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
ping @tianon

**Comment by calavera** [comment]
LGTM
